### ISSUE TITLE
Exposed schemas based on role `search_path`

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#2783]
## Problem

`db-schemas` is too rigid and PostgREST specific. Also adding more schemas dynamically currently requires superuser. `ALTER ROLE authenticator SET pgrst.db_schemas TO 'schema1, schema2 ..'`.

## Solution

Base the exposed schemas on the `search_path` of the authenticator members. So if:

```sql
ALTER ROLE anon SET search_path TO 'world';
ALTER ROLE webuser SET search_path TO 'api';
```

We would expose both `world` and `api`. `world` being accessible to both and `api` only to `webuser`.

Now adding more schemas dynamically can simply be:

```sql
ALTER ROLE webuser SET search_path TO 'public, api, new_schema';
NOTIFY pgrst, 'reload config';
```

Which is a postgres native way.

---

Why not base this on the USAGE privilege of the schema? Because all roles have implicit access on `pg_catalog` and `public`, we would end up exposing them as well.

---

This should be feasible now that we have https://postgrest.org/en/stable/references/auth.html#impersonated-role-settings.

**Comment by steve-chavez** [comment]
To avoid a breaking change, we could turn on this way with:

```
db-schemas-search-path = true
```

We would still need the `db-extra-search-path` for extensions.

**Comment by mkleczek** [comment]
I am not really sure if it is a good idea to use search_path on authenticator for configuration as it might have unexpected effects on the system - setting search path might override the functions executed before set role (it is possible today but after this change setting search_path would be much more common).

To be honest I don't really understand why we are looking for imperative ways to pass configuration by (ab)using Pg mechanisms and do not consider purely declarative and flexible #2703 - that one would cover all **current** and **future** cases of changes in configuration while being backwards compatible.

**Comment by steve-chavez** [comment]
> setting search path might override the functions executed before set role

Hm, can you elaborate? I'm thinking of basing this around the `authenticator` search_path, I only expect it to be able to execute pg_catalog(which search_path doesn't affect) functions plus the new [pre-config](https://github.com/PostgREST/postgrest/pull/2803)(which can be schema qualified).

> To be honest I don't really understand why we are looking for imperative ways to pass configuration by (ab)using Pg mechanisms 

We have been using pg [tx-scoped settings](https://postgrest.org/en/stable/references/transactions.html#transaction-scoped-settings) for a long time now. I don't think we're abusing them? They've been pretty stable.

> and do not consider purely declarative and flexible https://github.com/PostgREST/postgrest/issues/2703

- Too error prone. Injecting a SQL query would imply we need to understand its return type . With the `pre-config` approach we don't have to worry about that.
- Having a SQL query inside the config file feels like we're opening for SQL templating. The same argument you make can be used to argue about exposing endpoints based on queries("we don't want to create a VIEW, let us expose an endpoint based on a SQL query in the config").

**Comment by mkleczek** [comment]
> Having a SQL query inside the config file feels like we're opening for SQL templating. The same argument you make can be used to argue about exposing endpoints based on queries("we don't want to create a VIEW, let us expose an endpoint based on a SQL query in the config"). This would defeat the whole purpose of database as single source of truth.

Fair enough.

How about the solution along the lines of db-pre-request that would allow to provide the name of a view providing configuration?

**Comment by steve-chavez** [comment]
>  How about the solution along the lines of db-pre-request that would allow to provide the name of a view providing configuration?

How would that work? The view would also do `set_config` like the pre-config function(https://github.com/PostgREST/postgrest/pull/2803#issue-1726683726)?

**Comment by mkleczek** [comment]
Right now config is read from the db using dbSettingsStatement - which in turn has an SQL query hardcoded. My initial proposal was to make it possible to configure this query (and the default would stay as is).
But it would be equally fine if the code would check for "db-config-view" property and if set - read the config using "SELECT key, value FROM [db-config-view]"

**Comment by steve-chavez** [comment]
>But it would be equally fine if the code would check for "db-config-view" property and if set - read the config using "SELECT key, value FROM [db-config-view]"

Why not use the new [pre-config function](https://github.com/PostgREST/postgrest/pull/2803#issue-1726683726)?

Not opposed to the idea but seems like we should prefer one or the other. For now I see these advantages for a function:

- Easier to execute conditional logic based on roles. So the same function can be used for multiple instances.
- Less privileges. A function only requires USAGE privilege on the schema while a view also requires SELECT privilege.
- We don't have to worry about a bad return type.

So why a pre-config view? Can you provide an example of how would such a view look like?

**Comment by mkleczek** [comment]
> > But it would be equally fine if the code would check for "db-config-view" property and if set - read the config using "SELECT key, value FROM [db-config-view]"
> 
> Why not use the new [pre-config function](https://github.com/PostgREST/postgrest/pull/2803#issue-1726683726)?
> 

Honestly - I wasn't aware of the new pre-config function. So it seems right now discussion is moot and  it is only a matter of taste (imperative vs declarative).

**Comment by mkleczek** [comment]
That said: there is one advantage of configurable config query - no need to create any Postgrest dedicated database objects. For me this is a really important feature as deployment does not require any database migrations.

**Comment by steve-chavez** [comment]
Yeah, but then again we require creating some database roles.

A [database template](https://www.postgresql.org/docs/current/manage-ag-templatedbs.html) can also avoid migrations to certain extent.

**Comment by mkleczek** [comment]
> Yeah, but then again we require creating some database roles.
> 

My use case is that installing Postgrest does not require any database roles (except pgrst authenticator user). It can be treated as a part of infrastructure (managed by a different team than the DBAs). Applications opt in to expose their schemas by granting roles to pgrst authenticator and by updating the list of schemas.

**Comment by steve-chavez** [comment]
> (managed by a different team than the DBAs). Applications opt in to expose their schemas by granting roles to pgrst authenticator and by updating the list of schemas.

That seems like a perfect use case for [Trusted Language Extensions](https://github.com/aws/pg_tle). We can have `postgrest-contrib` as TLE and it could have a `select postgrest.expose_schemas('{api,rest,etc}')` function(using pre-config underlyingly). That can be controlled by your non-DBA team.

**Comment by steve-chavez** [comment]
> The authenticator role is also doing the one doing the schema cache lookup, so it needs to know at this stage already, which schemas to look at.

Yeah, it makes a lot of sense to base it of the authenticator. Seems we'll have to keep `db-extra-search-path` though.
